nsqlookupd/registration_db.go

```diff
@@ -134,10 +140,29 @@ func (r *RegistrationDB) FindRegistrations(category string, key string, subkey s
 }
 
 func (r *RegistrationDB) FindProducers(category string, key string, subkey string) Producers {
+	r.cachedMutex.RLock()
+	cachedKey := fmt.Sprintf("%s:%s:%s", category, key, subkey)
+
+	if val, found := r.cachedFindProducersResults.Get(cachedKey); found {
+		r.cachedMutex.RUnlock()
+		return val.(Producers)
+	}
+
+	r.cachedMutex.RUnlock()
```

Comment:
https://github.com/patrickmn/go-cache#go-cache says "the cache can be safely used by multiple goroutines" so I don't think you need this lock at all

Reply:
This is mainly used to avoid multiple set operations happen.

Reply:
I see, but you still don't need the read lock for that, only the write lock (which could be a plain mutex) for the following section, where you check again if the result is in the cache, before running the  expensive loop.

That expensive loop, which is the main thing you are trying to avoid, is:

```go
	for k, producers := range r.registrationMap {
		if !k.IsMatch(category, key, subkey) {
			continue
		}
		for _, producer := range producers {
```

I think that there is probably a better way to structure the registration map, or maintain pre-computed results along side it or embedded in it, which are updated or invalidated whenever the registration map is updated, to get better performance with huge numbers of producers / topics / channels / whatever.

Reply:
According to the doc:
> (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys.

the item one said that `sync.Map` suites for a situation where only one written happens which is definitely not the truth for this nsqlookupd registeration map usage.

Reply:
> but you still don't need the read lock for that, only the write lock (which could be a plain mutex) for the following section, where you check again if the result is in the cache, before running the expensive loop.

@ploxiln Thanks, i understand. Will do.

Reply:
Despite being less than optimal for this use-case, a `sync.Map` can have entries inserted and updated concurrently while a query is iterating over the entire structure. So it *might* resolve the initial complaint about the write lock being starved.

I consider the addition of this cache to be a hacky work-around. It adds up to a 1 minute delay to updates, for all users. It may be OK for many consumers, but probably not all, and not for nsqd or nsqadmin.

nsqlookupd is already just a big in-memory data structure. It could probably be re-structured to process your extreme case faster. That would admittedly require more time and effort to understand the current structure and queries. So a cheap quick fix, without those problems, *might* be a sync.Map.